pyop2/codegen/rep2loopy.py

```diff
@@ -85,6 +91,210 @@ def petsc_function_lookup(target, identifier):
+
+
+def inv_fn_lookup(target, identifier):
+    if identifier == 'inv':
+        return INVCallable(name='inv')
+    else:
+        return None
+
+
+class SolveCallable(loopy.ScalarCallable):
+    def __init__(self, name, arg_id_to_dtype=None,
+                 arg_id_to_descr=None, name_in_target=None):
+
+        super(SolveCallable, self).__init__(name,
+                                            arg_id_to_dtype=arg_id_to_dtype,
+                                            arg_id_to_descr=arg_id_to_descr)
+
+        self.name = name
+        self.name_in_target = name_in_target
+
+    def with_types(self, arg_id_to_dtype, kernel, callables_table):
+        for i in range(len(arg_id_to_dtype)):
+            if arg_id_to_dtype.get(i) is None:
+                # the types provided aren't mature enough to specialize the
+                # callable
+                return (self.copy(arg_id_to_dtype=arg_id_to_dtype),
+                        callables_table)
+
+        mat_dtype = arg_id_to_dtype[0].numpy_dtype
+        name_in_target = "solve"
+
+        return (self.copy(name_in_target=name_in_target,
+                arg_id_to_dtype={-1: NumpyType(mat_dtype, target=loopy.CTarget()), 0: NumpyType(mat_dtype, target=loopy.CTarget()), 1: NumpyType(mat_dtype, target=loopy.CTarget()), 2: NumpyType(numpy.int32, target=loopy.CTarget())}),
+                callables_table)
+
+    def emit_call_insn(self, insn, target, expression_to_code_mapper):
+        assert self.is_ready_for_codegen()
+        assert isinstance(insn, loopy.CallInstruction)
+
+        parameters = insn.expression.parameters
+
+        if type(parameters) != list:
+            parameters = list(parameters)
+        par_dtypes = [self.arg_id_to_dtype[i] for i, _ in enumerate(parameters)]
+
+        parameters.append(insn.assignees[0])
+        par_dtypes.append(self.arg_id_to_dtype[1])
+        mat_descr_A = self.arg_id_to_descr[0]
+
+        arg_c_parameters = [
+            expression_to_code_mapper(
+                par,
+                PREC_NONE,
+                dtype_to_type_context(target, par_dtype),
+                par_dtype
+            ).expr
+            if isinstance(par, SubArrayRef) else
+            expression_to_code_mapper(
+                par,
+                PREC_NONE,
+                dtype_to_type_context(target, par_dtype),
+                par_dtype
+            ).expr
+            for par, par_dtype in zip(parameters, par_dtypes)
+        ]
+        c_parameters = []
+        c_parameters.insert(0, arg_c_parameters[-1])  # out
+        c_parameters.insert(1, arg_c_parameters[0])  # A
+        c_parameters.insert(2, arg_c_parameters[1])  # B
+        c_parameters.insert(3, mat_descr_A.shape[1])  # n
+        return var(self.name_in_target)(*c_parameters), False
+
+    def generate_preambles(self, target):
+        assert isinstance(target, CTarget)
+        code = """#include <string.h>
+            #include <stdio.h>
+            #include <stdlib.h>
```

Comment:
Rather than putting these includes here, shall we just whack them in the standard set of headers we require?

Reply:
With the standard set of headers being defined in PyOP2 I assume? I don't think that will work until the inlining is done right. If we don't include the headers here, we will need to attach them to the slate kernel before the C device code (which is then passed to PyOP2) is generated from it, I believe. Would that be better? Then we might have headers in our kernels, which we do not always use however.

Reply:
I think we put the headers the kernel advertises in the preamble (see rep2loopy.py line 480 or so)? But this is probably fine.

Reply:
Yes we do, but I think that one is only called on Pyop2 level and I need the preambles already on slate level for the workaround of compiling the slate kernel to C code (before passing back to pyop2). I have moved the preamble to the slate compiler now, so we don't need to change the callables after we fixed the kernel inlining.